### PR TITLE
First cut at clipping

### PR DESCRIPTION
**Pull request by raphlinus**
This is a significant amount of progress towards #36. Bounding boxes for clips are accumulated (in world coordinates, see that issue for more details) CPU-side and encoded in the BeginClip and EndClip elements. These are carried through the pipeline in a straightforward way, then there is a blend stack in the fine rasterizer.

Here are the cleanups not yet done:

* The blend stack has finite size. Deal with the overflow (by writing to a scratch buffer in global memory)
* Clean up the older FillMask mechanism.
* Optimize all-clear and all-solid clip tiles.

This PR also encodes transforms (they have always worked, just weren't encoded).

### CODE REVIEW

File: piet-gpu/shader/kernel4.comp [hunk 1]
```diff
@@ -27,6 +27,8 @@ layout(rgba8, set = 0, binding = 2) uniform writeonly image2D image;
 #include "ptcl.h"
 #include "tile.h"
 
+#define BLEND_STACK_SIZE 4
```
Comment (eliasnaur) [inline piet-gpu/shader/kernel4.comp:30]: Will you lift this restriction before replacing FillMask?

Comment (raphlinus) [reply piet-gpu/shader/kernel4.comp:30]: I think so, but I'm not 100% sure when. On my AMD 5700 XT card, the static limit can be increased quite a bit without seriously impacting performance (utilization goes down as register pressure goes up). Using a scratch buffer in global memory can be done in two stages - malloc-only (using an atomic bump counter, as is currently done for dynamic allocations), and a second malloc/free stage.

Comment (eliasnaur) [reply piet-gpu/shader/kernel4.comp:30]: Can the FillMask stopgap co-exist with the new clipping? I'd hoped not to have hard limits such as this when adding piet-gpu as a Gio backend. I'll probably take a stab at #40, but I think this limit is worse.

File: piet-gpu/shader/coarse.comp [hunk 2]
```diff
@@ -224,20 +220,23 @@ void main() {
...
             uint width = sh_tile_width[el_ix];
             uint x = sh_tile_x0[el_ix] + seq_ix % width;
             uint y = sh_tile_y0[el_ix] + seq_ix / width;
-            Tile tile = Tile_read(TileRef(sh_tile_base[el_ix] + (sh_tile_stride[el_ix] * y + x) * Tile_size));
-            // Include the path in the tile if
-            // - the tile contains at least a segment (tile offset non-zero)
-            // - the tile is completely covered (backdrop non-zero)
-            // - the tile is not covered and we're filling everything outside the path (backdrop zero, inverse fills).
-            bool inside = tile.backdrop != 0;
-            bool fill = tag != Annotated_FillMaskInv;
```
Comment (eliasnaur) [inline piet-gpu/shader/coarse.comp:239]: I think this check needs restoring.

File: piet-gpu/shader/kernel4.comp (hunk 1 above)
Comment (eliasnaur) [reply piet-gpu/shader/kernel4.comp:30]: Thank you very much. If you decide to leave it, I'll take a look after #23, #38, #40 after which Gio has all it needs for now. The existing Gio renderer tests can be re-used and expanded to cover piet-gpu, and I'll be much more comfortable working on enhancements and bugs.

**Comment by eliasnaur** [comment]
Confirmed that FillMask still works with this branch.

**Comment by raphlinus** [comment]
Good to know. I'll take that as a go-ahead to commit when I feel it's appropriate. (FYI, generally the strategy I use is that my local branch is one commit ahead. When I'm ready to PR from that, I merge the one behind it into master). I've started the scratch buffer (malloc-only) and do not think it will be very hard. I'll also comment on the dynamic allocation issue.

**Pull request merged**